meteor/diffmaps.py

```diff
@@ -0,0 +1,195 @@
+
+    if not inplace:
+        dataset = dataset.copy()
+
+    # Convert native and derivative amplitude/phase pairs to complex arrays
+    native_complex = _rs_dataseries_to_complex_array(
+        dataset[native_amplitudes], dataset[native_phases]
+    )
+
+    if derivative_phases is not None:
+        derivative_complex = _rs_dataseries_to_complex_array(
+            dataset[derivative_amplitudes], dataset[derivative_phases]
+        )
+    else:
+        # If no derivative phases are provided, assume they are the same as native phases
+        derivative_complex = _rs_dataseries_to_complex_array(
+            dataset[derivative_amplitudes], dataset[native_phases]
+        )
+
+    # Compute complex differences
+    delta_complex = derivative_complex - native_complex
+
+    # Convert back to amplitude and phase DataSeries
+    delta_amplitudes, delta_phases = _complex_array_to_rs_dataseries(
+        delta_complex, dataset.index
+    )
+
+    # Add results to dataset
+    dataset["DeltaFoFo"] = delta_amplitudes
+    dataset["DeltaPhases"] = delta_phases
+
+    # Reset to Phase dtype
+    dataset["DeltaPhases"] = dataset["DeltaPhases"].astype("Phase")
+
+    if inplace:
+        return None
+    else:
+        return dataset
+
+
+def compute_kweights(
+    df: rs.DataSeries, sigdf: rs.DataSeries, kweight: float
+) -> rs.DataSeries:
+    """
+    Compute weights for each structure factor based on DeltaF and its uncertainty.
+    """
+    w = 1 + (sigdf**2 / (sigdf**2).mean()) + kweight * (df**2 / (df**2).mean())
+    return w**-1
+
+
+def compute_kweighted_deltafofo(
+    dataset: rs.DataSet,
+    *,
+    native_amplitudes: str,
+    derivative_amplitudes: str,
+    native_phases: str,
+    derivative_phases: str | None = None,
+    sigf_native: str,
+    sigf_deriv: str,
+    kweight: float = 1.0,
+    optimize_kweight: bool = False,
+    inplace: bool = False,
+) -> rs.DataSet | None:
+    """
+    Compute k-weighted differences between native and derivative amplitudes and phases.
+
+    Assumes that scaling has already been applied to the amplitudes before calling this function.
+
+    Parameters:
+    -----------
+    dataset : rs.DataSet
+        The input dataset containing columns for native and derivative amplitudes/phases.
+    native_amplitudes : str
+        Column label for native amplitudes in the dataset.
+    derivative_amplitudes : str
+        Column label for derivative amplitudes in the dataset.
+    native_phases : str, optional
+        Column label for native phases.
+    derivative_phases : str, optional
+        Column label for derivative phases, by default None.
+    sigf_native : str
+        Column label for uncertainties of native amplitudes.
+    sigf_deriv : str
+        Column label for uncertainties of derivative amplitudes.
+    kweight : float, optional
+        k-weight factor, by default 1.0.
+    optimize_kweight : bool, optional
+        Whether to optimize the kweight using negentropy, by default False.
```

Comment:
Alternative.

Just always provide an explicit list of `kweights_to_evaluate`. Defaults to `np.linspace(0.0, 1.0, 101)` ie `0.00, 0.01, 0.02, ..., 1.00`, but you could always just pass `[my_favorite_kweight]`.

Reply:
It's fine, I agree. Either user specifies favorite kweight, or enables optimization. Would be my vote.